src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java

```diff
@@ -452,32 +453,72 @@ else if (resource instanceof LayerInfo)
         return summaries;
     }
 
-    private void setRuleFilterUserOrRole(Authentication user, RuleFilter ruleFilter) {
+    private void setRuleFilterUserAndRole(Authentication user, RuleFilter ruleFilter) {
         if (user != null) {
             GeoFenceConfiguration config = configurationManager.getConfiguration();
-            if (config.isUseRolesToFilter() && config.getRoles().size() > 0) {
 
-                String role = "UNKNOWN";
+            // just some loggings here
+            if (config.isUseRolesToFilter()) {
+                if (config.getRoles().isEmpty()) {
+                    LOGGER.log(
+                            Level.WARNING,
+                            "Role filtering requested, but no roles provided. Will only use user authorizations");
+                }
+
+                if (LOGGER.isLoggable(Level.FINE)) {
+                    String authList =
+                            user.getAuthorities().stream()
+                                    .map(a -> a.getAuthority())
+                                    .collect(Collectors.joining(",", "[", "]"));
+                    LOGGER.log(
+                            Level.FINE,
+                            "Authorizations found for user {0}: {1}",
+                            new Object[] {user.getName(), authList});
+                }
+            }
+
+            if (config.isUseRolesToFilter() && !config.getRoles().isEmpty()) {
+
+                boolean getAllRoles = config.getRoles().contains("*");
+                Set<String> excluded =
+                        user.getAuthorities().stream()
+                                .map(a -> a.getAuthority())
+                                .filter(r -> r.startsWith("-"))
```

Comment:
It would be better to document this - prefix to exclude roles from being included in the filter.